src/Forms/HTMLEditor/HTMLEditorElementRule.php

```diff
@@ -0,0 +1,269 @@
+<?php
+
+namespace SilverStripe\Forms\HTMLEditor;
+
+use DOMAttr;
+use DOMElement;
+use InvalidArgumentException;
+use SilverStripe\Core\ArrayLib;
+
+/**
+ * Rules used to define whether a given element or regex pattern of elements is allowed in the content of an HTMLEditorField.
+ */
+class HTMLEditorElementRule
+{
+    /**
+     * Name of the global rule for a given rule set
+     */
+    public const string GLOBAL_NAME = '_global';
+
+    /**
+     * Name of the element(s) this rule applies to.
+     * This can also be a regex pattern - in which case nameIsPattern should be set to `true`.
+     */
+    private string $name;
+
+    /**
+     * Indicates whether name is a regex pattern to match against.
+     */
+    private bool $nameIsPattern = false;
+
+    /**
+     * Associative array of attribute rules which apply to this element rule.
+     * Excludes rules which use regex - see $attributePatternRules
+     * @var HTMLEditorAttributeRule[]
+     */
+    private array $attributeRules = [];
+
+    /**
+     * Associative array of attribute rules with regex expressions which apply to this element rule.
+     * See also $attributeRules
+     * @var HTMLEditorAttributeRule[]
+     */
+    private array $attributePatternRules = [];
+
+    /**
+     * Array of attribute rules defining attributes that must be present.
+     *
+     * If none of the attributes are present, the element(s) this rule applies to is not allowed.
+     * Note that only one of the attributes in this list has to be present for the element to be allowed.
+     * @var HTMLEditorAttributeRule[]
+     */
+    private array $requiredAttributeRules = [];
+
+    /**
+     * If true, empty elements will have a non-breaking space added inside them.
+     */
+    private bool $padEmpty = false;
+
+    /**
+     * If true, empty elements will be removed.
+     */
+    private bool $removeIfEmpty = false;
+
+    /**
+     * If true, elements with no attributes will be removed.
+     */
+    private bool $removeIfNoAttributes = false;
+
+    private ?HTMLEditorElementRule $globalRule = null;
+
+    /**
+     * Returns a HTMLEditorAttributeRule based on an associative array defining the rule.
+     * If the rule array is empty, the defaults apply.
+     * Excludes attributes - get those separately using HTMLEditorAttributeRule::fromArray().
+     */
+    public static function fromArray(string $name, array $ruleArray): static
```

Comment:
Move method below the constructor

Reply:
Discussion can be in https://github.com/silverstripe/silverstripe-framework/pull/11617#discussion_r1988366988